### PR TITLE
feat: SyncUserAction and SyncUserActionHandler - WPB-544

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-544" title="WPB-544" target="_blank">StoryWPB-544</a>  [iOS] Client periodically checks if migration can be finalised
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This `SyncUserAction` and the `SyncUserHandler` is needed for the work regarding migration. 

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift [hunk 1]
```diff
@@ -58,12 +62,11 @@ public class UserProfileRequestStrategy: AbstractRequestStrategy, IdentifierObje
...
-            return userProfileByID.nextRequest(for: apiVersion)
-        case .v1, .v2, .v3, .v4, .v5:
-            return userProfileByQualifiedID.nextRequest(for: apiVersion)
-        }
+
+        return
+            userProfileByID.nextRequest(for: apiVersion) ??
+            userProfileByQualifiedID.nextRequest(for: apiVersion) ??
+            actionSync.nextRequest(for: apiVersion)
     }
```
Comment (agisilaos) [inline wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]: @johnxnguyen Is this something that we can do regarding the requests? I would like the requests to execute one after the other or to make sure that `actionSync.nextRequest(for: apiVersion)` is being executed

File: wire-ios-data-model/Source/Model/Conversation/Actions/SyncUsersAction.swift [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        case invalidBody
+        case invalidResponsePayload
+        case endpointUnavailable
+        case failedToEncodeRequestPayload
+        case unknownError(code: Int, label: String, message: String)
+    }
+
+    // MARK: - Properties
+
+    public let qualifiedID: [QualifiedID]
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Model/Conversation/Actions/SyncUsersAction.swift]:
```suggestion
    public let qualifiedIDs: [QualifiedID]
```

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+
+// rename to describe v4
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]: What do you mean by this?

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        let qualified_user_map: Payload.UserProfiles
+
+    }
+
+    override func handleResponse(
+        _ response: ZMTransportResponse,
+        action: SyncUsersAction
+    ) {
+        guard let apiVersion = APIVersion(rawValue: response.apiVersion) else {
+            return
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift:94]: We should still fail the action in this case.

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        var action = action
+
+        switch apiVersion {
+        case .v0, .v1, .v2, .v3:
+            return
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift:100]: We should still fail the action in this case.

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            switch response.httpStatus {
+            case 200:
+                guard let rawData = response.rawData,
+                      let payload = Payload.UserProfilesV4(rawData)
+                else {
+                    action.fail(with: .invalidResponsePayload)
+                    return
+                }
+
+                var result: () = SyncUsersAction.Result()
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]: No need to initialize `Void`, just call `action.succeed()` (I think)

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+    }
+
+    private func markUserProfilesAsUnavailable(_ users: Set<QualifiedID>) {
+        for qualifiedID in users {
+            let user = ZMUser.fetch(with: qualifiedID.uuid, domain: qualifiedID.domain, in: context)
+            user?.isPendingMetadataRefresh = true
+            user?.needsToBeUpdatedFromBackend = false
+        }
+    }
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]: You should probably do this in `context.performAndWait`.

File: wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]: Yes but should should make sure that `userProfileByID` returns a `nil` when the api version is 1 or greater.

Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]: @johnxnguyen So something like this?

```swift
public override func nextRequestIfAllowed(for apiVersion: APIVersion) -> ZMTransportRequest? {
    fetchAllConnectedUsers(for: apiVersion)

    // Check the apiVersion before userProfileByID.nextRequest
    if apiVersion.rawValue >= 1 {
        return
            userProfileByQualifiedID.nextRequest(for: apiVersion) ??
            actionSync.nextRequest(for: apiVersion)
    } else {
        return
            userProfileByID.nextRequest(for: apiVersion) ??
            userProfileByQualifiedID.nextRequest(for: apiVersion) ??
            actionSync.nextRequest(for: apiVersion)
    }
}
```

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift (hunk 5 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift:100]: Done!

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift (hunk 6 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]: Indeed, you're right, I don't need it. Fixed.

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift (hunk 4 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift:94]: Done!

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift (hunk 3 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]: To rename the class to it's related to V4 and above. Do you think that would be a helpful change?

File: wire-ios-data-model/Source/Model/Conversation/Actions/SyncUsersAction.swift (hunk 2 above)
Comment (agisilaos) [reply wire-ios-data-model/Source/Model/Conversation/Actions/SyncUsersAction.swift]: Done!

File: wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]: I was thinking more like the `userProfileByIDTranscoder` returning nil if the api version is greater than zero.

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift (hunk 7 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]: Done!

File: wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift (hunk 1 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]: Fixed it, thank you!

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    func testRequestGenerationForAPIVersionV4() {
+        // GIVEN
+        let sut = SyncUsersActionHandler(context: uiMOC)
+        let id = QualifiedID(uuid: .create(), domain: "example.com")
+        let action = SyncUsersAction(qualifiedIDs: [id])
+
+        // WHEN
+        let result = try! XCTUnwrap(sut.request(for: action, apiVersion: .v4))
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift]: Mark the test with 'throws' and remove `!`.

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        XCTAssertNotNil(result, "Request should not be nil for API version .v5")
+        XCTAssertEqual(result.path, "/v5/list-users", "Incorrect path for API version .v5")
+        XCTAssertEqual(result.method, .post, "Incorrect HTTP method for API version .v5")
+        XCTAssertEqual(result.apiVersion, APIVersion.v5.rawValue, "Incorrect API version set in the request for .v5")
+
+        let payload = try! XCTUnwrap(RequestPayload(result))
+        XCTAssertEqual(payload, RequestPayload(qualified_users: [id]))
+    }
+
+    // MARK: - Response Handling
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift:74]: There should also be a test for the success.

File: wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift [hunk 10]
```diff
@@ -58,12 +62,11 @@ public class UserProfileRequestStrategy: AbstractRequestStrategy, IdentifierObje
...
-        case .v0:
-            return userProfileByID.nextRequest(for: apiVersion)
-        case .v1, .v2, .v3, .v4, .v5:
-            return userProfileByQualifiedID.nextRequest(for: apiVersion)
-        }
+
+        return
+        userProfileByID.nextRequest(for: apiVersion) ??
+        userProfileByQualifiedID.nextRequest(for: apiVersion) ??
+        actionSync.nextRequest(for: apiVersion)
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]: I think you can write this as

```swift
return [userProfileByID, userProfileByQualifiedID, actionSync].nextRequest(for: apiVersion)
```

File: wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift [hunk 11]
```diff
@@ -240,6 +243,10 @@ class UserProfileByIDTranscoder: IdentifierObjectSyncTranscoder {
     }
 
     func request(for identifiers: Set<UUID>, apiVersion: APIVersion) -> ZMTransportRequest? {
+        if apiVersion.rawValue > 0 {
+            return nil
+        }
+
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]:
```suggestion
        guard apiVersion == .v0 else { return nil }
```

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift (hunk 3 above)
Comment (johnxnguyen) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]: Can we also make the action support the other preview api versions too?

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift (hunk 8 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift]: Done!

File: wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift (hunk 11 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]: Fixed it!

File: wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift (hunk 10 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/User/UserProfileRequestStrategy.swift]: Yep, we can write it like this. I fixed it.

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+    // MARK: - Response
+
+    struct ResponsePayload: Codable {
+
+        let qualified_user_map: Payload.UserProfiles
+
+    }
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]:
```suggestion
```

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import XCTest
+@testable import WireRequestStrategy
+
+final class SyncUsersActionHandlerTests: MessagingTestBase {
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift]: Your tests can be simplified by using the helper methods from `ActionHandlerTestBase` if you make your test class a subclass of it.

See `SendMLSMessageActionHandlerTests` for examples of how to handle tests for request generation and response handling in success and failure scenarios

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift (hunk 9 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift:74]: Done! @johnxnguyen

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift (hunk 13 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandlerTests.swift]: Done!

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift (hunk 3 above)
Comment (agisilaos) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift]: This is out of the scope of this PR. I will create a ticket for it. I already discussed it with @David-Henner!

**Review by David-Henner — APPROVED** [review]

**Pull request merged**